cwl/src/main/scala/cwl/WorkflowStepInput.scala

```diff
@@ -0,0 +1,146 @@
+                       expressionLib: ExpressionLib,
+                       matchingRunInputType: Option[MyriadInputType],
+                       isScattered: Boolean
+                      )(implicit parentName: ParentName): ErrorOr[ExposedExpressionNode] = {
+
+    val inputs = sourceMappings.keySet
+
+    val identifier = WomIdentifier(id)
+
+    val node =
+      for {
+        inputType <- WorkflowStepInput.determineType(this, outputTypeMap, matchingRunInputType, isScattered)
+        womExpression = WorkflowStepInputExpression(this, inputType, inputs, expressionLib)
+        node <- ExposedExpressionNode.fromInputMapping(identifier, womExpression, inputType, sourceMappings).toEither
+      } yield node
+    node.toValidated
+  }
+
+  lazy val sources: List[String] =
+    source.
+      toList.
+      flatMap(_.fold(StringOrStringArrayToStringList))
+
+  lazy val effectiveLinkMerge: LinkMergeMethod = linkMerge.getOrElse(LinkMergeMethod.MergeNested)
+
+  def sourceValues(inputValues: Map[String, WomValue])(implicit pn: ParentName): Checked[Map[String, WomValue]]  = {
+
+    def lookupValue(key: String): Checked[WomValue] =
+      inputValues.
+        get(FullyQualifiedName(key).id).
+        toRight(s"source value $key not found in input values ${inputValues.mkString("\n")}." |> NonEmptyList.one)
+
+
+    sources.
+      traverse[ErrorOr, (String, WomValue)](s => lookupValue(s).toValidated.map(FullyQualifiedName(s).id -> _)).
+      toEither.
+      map(_.toMap)
+  }
+
+  def validatedSourceTypes(typeMap: WomTypeMap)(implicit pn: ParentName): Checked[WomTypeMap] = {
+    def lookupValue(key: String): Checked[WomType] =
+      typeMap.
+        get(FullyQualifiedName(key).id).
+        toRight(s"source value $key not found in type map ${typeMap.mkString("\n")}." |> NonEmptyList.one)
+
+    sources.
+      traverse[ErrorOr, (String, WomType)](s => lookupValue(s).toValidated.map(FullyQualifiedName(s).id -> _)).
+      toEither.
+      map(_.toMap)
+  }
+
+}
+
+object WorkflowStepInput {
+  type InputSource = String :+: Array[String] :+: CNil
+
+  def determineType(stepInput: WorkflowStepInput,
+                    outputTypeMap: Map[String, WomType],
+                    expectedType: Option[MyriadInputType],
+                    isScattered: Boolean)(implicit parentName: ParentName): Checked[WomType] = {
+
+    val expectedTypeAsWom: Option[WomType] = expectedType.map(_.fold(MyriadInputTypeToWomType))
+    (isScattered, expectedTypeAsWom, stepInput.effectiveLinkMerge) match {
+
+      //If scattering over this variable, we expect an array of the sink type
+      case (true, Some(tpe), _) => WomArrayType(tpe).asRight
+
+      //If sink parameter is an array, we must frame the input as an array
+      case (false, Some(array: WomArrayType), LinkMergeMethod.MergeNested) =>
+        array.asRight
+
+      //If sink parameter is an array and merge_flattened is used, must validate input & output types are equivalent before proceeding
```

Comment:
why is there compatibility checking in this case but not in other cases?

Reply:
In this case, it's in the spec in the merge nested section:

>The source and sink parameters must be compatible types

I think there is a gap in the logic of this requirement which is what if there is an expression in `valueFrom`?  I believe it's not possible to predict what type will come out of the expression,  and in any case compatibility between sources going into the expression and the ultimate input type are irrelevant.   In any case, the spec does not address this issue.

I guess in general I think we'd be fine to assert compatibility across all cases where `valueFrom` == `None`, but it's kinda weird that the spec doesn't say that in general.  (I also remember some catchall clause where if the value produced doesn't match the expected type, then try to proceed anyways).